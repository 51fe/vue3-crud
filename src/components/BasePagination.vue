<script setup lang="ts">
import { computed } from 'vue'
import { scrollTo } from '../utils/scroll-to'

const props = withDefaults(
  defineProps<{
    total: number
    page: number
    limit: number
    layout?: string
    background?: boolean
    autoScroll?: boolean
    hidden?: boolean
  }>(),
  {
    total: 0,
    page: 1,
    limit: 15,
    layout: 'total, sizes, prev, pager, next, jumper',
    autoScroll: true
  }
)

const emit = defineEmits<{
  (e: 'update:page', value: number): void
  (e: 'update:limit', value: number): void
  (e: 'pagination', value: number): void
}>()

const currentPage = computed({
  get: () => props.page,
  set: (val) => {
    emit('update:page', val)
  }
})

const pageSize = computed({
  get: () => props.limit,
  set: (val) => {
    emit('update:limit', val)
  }
})

function handleSizeChange(val: number) {
  emit('pagination', val)
  if (props.autoScroll) {
    scrollTo(0, 800)
  }
}
function handleCurrentChange(val: number) {
  emit('pagination', val)
  if (props.autoScroll) {
    scrollTo(0, 800)
  }
}
</script>

<template>
  <div
    :class="{ hidden: hidden }"
    class="pagination-container"
  >
    <el-pagination
      v-model:currentPage="currentPage"
      v-model:pageSize="pageSize"
      :layout="layout"
      :page-sizes="[15, 30, 50, 100]"
      :total="total"
      :background="background"
      @size-change="handleSizeChange"
      @current-change="handleCurrentChange"
    />
  </div>
</template>

<style lang="scss">
.pagination-container {
  display: flex;
  clear: both;
  background: #fff;
  .el-pagination {
    margin: 20px auto;
  }
  .hidden {
    display: none;
  }
}
</style>
