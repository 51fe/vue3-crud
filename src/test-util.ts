import { render, RenderOptions, RenderResult } from '@testing-library/vue'

import ElementPlus from 'element-plus'
import CN from 'element-plus/lib/locale/lang/zh-cn'

export const customRender = (
  ui: any, overrides?: Omit<RenderOptions, 'global'>): RenderResult => {
  return render(ui, {
    global: {
      plugins: [[ElementPlus, ...[{ locale:  CN }]]]
    },
    ...overrides
  })
}
export { defineComponent, ref, reactive, nextTick } from 'vue'
export * from '@testing-library/vue'
export { customRender as render }