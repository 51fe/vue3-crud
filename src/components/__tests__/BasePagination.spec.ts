import { fireEvent, screen, render } from '../../test-util'
import BasePagination from '../BasePagination.vue'

const setUp = () => {
  return render(BasePagination, {
    props: {
      total: 32
    }
  })
}

test('renders correctly by default', async () => {
  const { html } = setUp()
  screen.getByText(/共 32 条/i)
  expect(html()).toMatchSnapshot()
})

test('emits events correctly when click the pager number', async () => {
  const page = 2
  const { emitted } = setUp()
  const numberBtn = screen.getByText(page, { selector: '.number' })
  // click page 2 button
  await fireEvent.click(numberBtn)
  expect(emitted()['update:page'][0]).toEqual([page])
  expect(emitted()['pagination'][0]).toEqual([page])
})

test('emits events correctly when click the page size item', async () => {
  const limit = 30
  const { emitted } = setUp()
  const item = screen.getByText(new RegExp(`${limit}条/页`))
  // click the second item
  await fireEvent.click(item)
  expect(emitted()['update:limit'][0]).toEqual([limit])
  expect(emitted()['pagination'][0]).toEqual([limit])
})
