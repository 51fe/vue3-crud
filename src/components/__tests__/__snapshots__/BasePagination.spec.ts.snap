// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
"<div class="pagination-container">
  <div role="pagination" aria-label="pagination" class="el-pagination"><span class="el-pagination__total is-first" disabled="false">共 32 条</span><span class="el-pagination__sizes"><div class="el-select el-select--default"><div class="select-trigger el-tooltip__trigger el-tooltip__trigger"><!--v-if--><div class="el-input el-input--default el-input--suffix"><!-- input --><!-- prepend slot --><!--v-if--><div class="el-input__wrapper"><!-- prefix slot --><!--v-if--><input class="el-input__inner" type="text" readonly="" autocomplete="off" tabindex="0" placeholder="请选择"><!-- suffix slot --><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-icon el-select__caret el-select__icon"><svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg"><path fill="currentColor" d="M831.872 340.864 512 652.672 192.128 340.864a30.592 30.592 0 0 0-42.752 0 29.12 29.12 0 0 0 0 41.6L489.664 714.24a32 32 0 0 0 44.672 0l340.288-331.712a29.12 29.12 0 0 0 0-41.728 30.592 30.592 0 0 0-42.752 0z"></path></svg></i><!--v-if--><!--v-if--><!--v-if--><!--v-if--><!--v-if--><!--v-if--></span></span></div><!-- append slot -->
  <!--v-if-->
</div>
</div>
<!--teleport start-->
<!--teleport end-->
</div></span><button type="button" class="btn-prev" disabled="" aria-disabled="true"><i class="el-icon"><svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
      <path fill="currentColor" d="M609.408 149.376 277.76 489.6a32 32 0 0 0 0 44.672l331.648 340.352a29.12 29.12 0 0 0 41.728 0 30.592 30.592 0 0 0 0-42.752L339.264 511.936l311.872-319.872a30.592 30.592 0 0 0 0-42.688 29.12 29.12 0 0 0-41.728 0z"></path>
    </svg></i></button>
<ul class="el-pager">
  <li class="is-active number" aria-current="true" tabindex="0"> 1 </li>
  <!--v-if-->
  <li class="number" aria-current="false" tabindex="0">2</li>
  <!--v-if-->
  <li class="number" aria-current="false" tabindex="0">3</li>
</ul><button type="button" class="btn-next" aria-disabled="false"><i class="el-icon"><svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
      <path fill="currentColor" d="M340.864 149.312a30.592 30.592 0 0 0 0 42.752L652.736 512 340.864 831.872a30.592 30.592 0 0 0 0 42.752 29.12 29.12 0 0 0 41.728 0L714.24 534.336a32 32 0 0 0 0-44.672L382.592 149.376a29.12 29.12 0 0 0-41.728 0z"></path>
    </svg></i></button><span class="el-pagination__jump is-last" disabled="false">前往 <div class="el-input el-input--small el-pagination__editor is-in-pagination"><!-- input --><!-- prepend slot --><!--v-if--><div class="el-input__wrapper"><!-- prefix slot --><!--v-if--><input class="el-input__inner" min="1" max="3" type="number" autocomplete="off" tabindex="0"><!-- suffix slot --><!--v-if--></div><!-- append slot --><!--v-if--></div> 页</span></div>
</div>"
`;
